### PR TITLE
Add virtual scrolling for CourseList

### LINKED ISSUE #137: Bug — Course Search Input Delay
**Describe the bug**
When I search for a course name in the course list, there is a noticeable impact on the performance of the component. If you keep backspacing an entered course name, the keystrokes are buffered until it finishes 'undoing' the filter. 

**To Reproduce**
1. Have Summer 2020 loaded & go to Schedule page
2. Type, for example, "Com"
3. Backspace 3 times. Notice how there is a delay in input when you reach "Co"

**Expected behavior**
No noticeable delay.

**Screenshots**
searchDelay

**Desktop (please complete the following information):**
 - OS: Win 10
 - Browser: Chrome
 - Version: 80.0.3987.163 64-bit

**Additional context**
Could only apply the filter if there's > 2 characters?

### LINKED ISSUE #144: Minor Issues Clicking Box to Add a Course (Lag/ Click box Range)
**Describe the bug**

**To Reproduce**
1. Go to Course Schedule
2. Click on multiple classes in succession as if adding them all to your schedule
4. (Lag) At least from my device, I'm not sure if it is my laptop or the site in general, but it takes a bit for each class to load. It doesn't show it clicked for a bit, so I re-click it and accidentally un-selects it.
5. (Click box) while I was clicking classes, not all of the ones I clicked on went onto the schedule. I realized it was because the space above and below the name of the class isn't clickable space. 

**Expected behavior**
Less lag and to be able to click anywhere within the boundary of the lines to select the course. 

**Additional context**
I'm not sure if these are intentional things that are supposed to be like that, but I just noticed it was kind of slow and it was kind of annoying to select things quickly.

### LINKED ISSUE #148: Feature - Virtual Scrolling
**Job**
Performance issues surrounding course list rendering.

**Describe a possible solution**
Virtual scrolling
Only render a small number of courses at a time. 
Hopefully this can mitigate performance issues surrounding the large number of rerenders. 

**Context**
For RPI, a normal semester contains around a thousand courses. The current implementation holds all the courses in a list with each course being its own object. Thus every course is rendered on component mount and any change involves rerendering the entire list. It is my understanding that due to limitations in Vue 2, changing properties of objects in a list causes the entire list to be rerendered.

### PR DESCRIPTION
**Pull request by 123joshuawu**
**Issue**

Closes #137
Closes #144 
Closes #148

**Database Changes/Migrations**

N/A

**Test Modifications**

N/A

**Test Procedure**

The performance photos below are using the following test procedure on the Fall 2020 data:
1) Add first course in CourseList
2) Switch to SelectedCourses view
3) Unselect selected course section
4) Switch to CourseList view
5) Remove selected course

**Photos**

Vue Devtools Performance Before Virtual Scrolling
1a)
Screen Shot 2020-06-03 at 11 04 21 PM
1b)
Screen Shot 2020-06-03 at 11 04 02 PM

After Virtual Scrolling
2a)
Screen Shot 2020-06-03 at 11 19 35 PM
2b)
Screen Shot 2020-06-03 at 11 19 28 PM

**Additional Info**

a) images show the total amount of time spent on each component (including time spent on each phase of the component lifecycle)
b) images show the browser fps with the `E` circles representing events

For my testing, my max fps is ~60 (blue bars). In 1a) all events (adding/removing course/sections) cause fps drops to around ~10 (red bars). After adding virtual scroll, the lowest fps reached is ~40. The total time spent on the components also dropped drastically.

Basically, the virtual scroll made everything zippy again.

### CODE REVIEW

**Comment by CoreyBuckley** [comment]
This is very good! Overall performance is much faster and is a much better experience. Course filtering is fast and my issue brought up in #137 seems non-reproducible.

I'm glad the v-scroll had such a dramatic effect.
(I like the generation of the _Requirements:_ string as well.)

Great job :)

**Comment by jshom** [comment]
This. Is. Awesome
